sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusProcessorClient.java

```diff
@@ -0,0 +1,216 @@
+     */
+    public synchronized void close() {
+        isRunning.set(false);
+        if (receiverSubscription.get() != null) {
+            receiverSubscription.get().cancel();
+        }
+        asyncClient.get().close();
+        scheduledExecutor.shutdown();
+    }
+
+    /**
+     * Returns {@code true} if the processor is running. If the processor is stopped or closed, this method returns
+     * {@code false}.
+     *
+     * @return {@code true} if the processor is running.
+     */
+    public synchronized boolean isRunning() {
+        return isRunning.get();
+    }
+
+    private synchronized void receiveMessages() {
+        if (receiverSubscription.get() != null) {
+            receiverSubscription.get().request(1);
+            return;
+        }
+        ServiceBusReceiverAsyncClient receiverClient = asyncClient.get();
+        receiverClient.receiveMessages()
+            .parallel(processorOptions.getMaxConcurrentCalls())
+            .runOn(Schedulers.boundedElastic())
+            .subscribe(new Subscriber<ServiceBusReceivedMessageContext>() {
+                @Override
+                public void onSubscribe(Subscription subscription) {
+                    receiverSubscription.set(subscription);
+                    receiverSubscription.get().request(1);
+                }
+
+                @Override
+                public void onNext(ServiceBusReceivedMessageContext serviceBusReceivedMessageContext) {
+                    if (serviceBusReceivedMessageContext.hasError()) {
+                        handleError(serviceBusReceivedMessageContext.getThrowable());
+                    } else {
+                        try {
+                            ServiceBusProcessorMessageContext serviceBusProcessorMessageContext =
+                                new ServiceBusProcessorMessageContext(receiverClient, serviceBusReceivedMessageContext);
+                            processMessage.accept(serviceBusProcessorMessageContext);
+                        } catch (Exception exception) {
+                            handleError(exception);
+                        }
+                    }
+                    if (isRunning.get()) {
+                        logger.verbose("Requesting 1 more message from upstream");
+                        receiverSubscription.get().request(1);
+                    }
+                }
+
+                @Override
+                public void onError(Throwable throwable) {
+                    logger.info("Error receiving messages.", throwable);
+                    handleError(throwable);
+                    if (isRunning.get()) {
+                        restartMessageReceiver();
+                    }
+                }
+
+                @Override
+                public void onComplete() {
+                    logger.info("Completed receiving messages.");
+                    if (isRunning.get()) {
+                        restartMessageReceiver();
```

Comment:
receiveMessages calls restartMessageReceiver, which will call receiveMessages again. Will this cause stackoverflow if the processor keeps running for long enough time?

Reply:
I wrote a code snippet to test it. It doesn't. But It does throw stackoverflow if I comment out two lines (parallel and runOn). You're safe because you use multiple threads.
```java
import reactor.core.publisher.Flux;
import reactor.core.scheduler.Schedulers;
import java.io.IOException;

public class Main {
    public static void foo() {
        Flux.just(1, 2)
                .parallel(2)
                .runOn(Schedulers.boundedElastic())
                .subscribe(
                        System.out::println,
                        System.out::println,
                        Main::foo2
        );
    }
    public static void foo2() {
        foo();
    }
    public static void main(String[] args) throws IOException {
        foo();
        System.in.read();
    }
}
```

Reply:
Yeah, this will run on a different thread pool and shouldn't cause stackoverflow.